runtime/minor_gc.c

```diff
@@ -641,13 +644,32 @@ void caml_empty_minor_heap_promote(caml_domain_state* domain,
     + (domain->young_end - domain->young_start) / 2;
   caml_reset_young_limit(domain);
 
+  domain->stat_minor_words += Wsize_bsize (minor_allocated_bytes);
+  domain->stat_promoted_words += domain->allocated_words - prev_alloc_words;
+
+  /* gc stats may be accessed unsynchronised by mutator code, so we collect the
+     sample before arriving at the barrier, which ensures that it doesn't race
+  */
+  caml_collect_gc_stats_sample(domain);
+
+  /* The code above is synchronised with other domains by the barrier below,
+     which is split into two steps, "arriving" and "leaving". When the final
+     domain arrives at the barrier, all other domains are free to leave, after
+     which they finish running the STW callback and may, depending on the
+     specific STW section, begin executing mutator code.
+
+     Leaving the barrier synchronises (only) with the arrivals of other domains,
+     so that all writes performed by a domain before arrival "happen-before" any
+     domain leaves the barrier. However, any code after arrival, including the
+     code between the two steps, can potentially race with mutator code.
+  */
+
+  /* arrive at the barrier */
   if( participating_count > 1 ) {
     atomic_fetch_add_explicit
-      (&domains_finished_minor_gc, 1, memory_order_release);
+      (&domains_finished_minor_gc, 1, memory_order_acq_rel);
```

Comment:
Overall the code seems fine to me as we discussed, but I am no memory-model expert and I cannot approve a change of memory ordering in this critical part of the code. I think that you could/should leave it as it is, and possibly open a separate issue or PR for it.

Reply:
You are correct, actually, I don't need to lift the atomic. It's a miserable `lock incq` or `lock addq` on x86 anyway... It took me reading the C11 standard to reassure myself. Namely, the relevant points in section 5.1.2.4 (Multi-threaded executions and data races) are:

10. A _release sequence_ headed by a release operation A on an atomic object M is a maximal
contiguous sub-sequence of side effects in the modification order of M, where the first
operation is A and every subsequent operation either is performed by the same thread that
performed the release or is an atomic read-modify-write operation.
11. Certain library calls _synchronize_ with other library calls performed by another thread. In
particular, an atomic operation A that performs a release operation on an object M
synchronizes with an atomic operation B that performs an acquire operation on M and
reads a value written by any side effect in the release sequence headed by A.

So domains `acquire`-ing the barrier variable at the end is enough for them to synchronise with _all_ increments, because even the first-arriving thread's `release` is properly synchronised with by virtue of the release sequence headed by it.